### DISCUSSION TITLE
Clustering

### DISCUSSION THREAD

**Discussion by joejnt** [teamalgoritma/community#773] [Q&A] [answered]
Dear rekan2 algoritma,

Selamat sore,
Mohon sarannya, saya sedang membuat clustering dengan syntax sebagai berikut:
set.seed(100)
pca_km <- eclust(df_pca,stand = TRUE,FUNcluster = "kmeans",k = 4, graph = T)

(sebelumnya saya sudah menerapkan pca terlebih dahulu untuk mengurangi dimensi), kemudian saya ingin mengukur kualitas klaster dengan menggunakan syntax berikut:

fviz_nbclust(df_pca, kmeans, method = "silhouette") +
  labs(subtitle = "Silhouette method") dan 

fviz_nbclust(df_pca, kmeans, method = "wss") +
  geom_line(xintercept = 4, linetype = 2) +
  labs(subtitle = "Elbow Method")
  
Namun ternyata komputasinya cukup lama saya sudah jalankan dari pukul 08.00 sampai dengan sore hari tidak selesai, apakah ada cara supaya bisa mengukur kualitas kluster tanpa membuat visualisasi.

Adapun jumlah obs data sebanyak 25,437 dan 3 var

Tambahan:
untuk mengukur kualitas cluster dengan elbow method, saya coba menggunakan syntax ini:
wss_values_pca <- numeric(length(k_values))

# Loop untuk menghitung WSS untuk setiap k
for (k in k_values) {
  kmeans_model <- kmeans(df_pca, centers = k)
  wss_values_pca[k] <- sum(kmeans_model$tot.withinss)
}

# Menampilkan nilai WSS untuk setiap k
print(wss_values_pca)

# Membuat data frame
wss_df_pca <- data.frame(K = 1:length(wss_values_pca), 
                     WSS = wss_values_pca)
wss_df_pca

hasilnya seperti ini:
<img width="194" alt="image" src="https://github.com/teamalgoritma/community/assets/103427989/4bb85442-63e3-46ed-afe2-6f05798b3900">

Tapi masih sedikit ragu, mohon sarannya.

Best regards,
JT

**Comment by ramdhinafinita** [comment]
Dear @joejnt 
Terimakasih atas pertanyaannya.  Berikut ini saran yang bisa saya berikan:
Untuk mengukur kualitas cluster tanpa harus membuat visualisasi, Anda bisa menggunakan metrik evaluasi kualitas klaster seperti indeks Silhouette atau Within-Cluster-Sum-of-Squares (WSS) langsung dalam kode R. Untuk yang bagian perhitungan WSS sudah dilakukan dengan benar dalam contoh yang Anda berikan. Anda bisa melanjutkan dengan memeriksa nilai WSS untuk berbagai nilai k (jumlah klaster) yang berbeda. Anda dapat menemukan nilai optimal k dengan melihat di mana kurva WSS mulai merata (elbow point).
Nilai WSS yang bagus itu jika di visualisasikan, mencari titik di mana penurunan WSS mulai melambat dan membentuk lengkungan yang mirip dengan "siku" pada plot. Dari hasil data bapak kurang lebih plot Elbow nya seperti ini

![image](https://github.com/teamalgoritma/community/assets/64061969/b3a04797-4024-4d73-90cf-0381ee4fca9a)
Sehingga bisa dilihat yang mulai melandai dan membentuk siku adalah di K 5. Namun, ini adalah pendekatan yang sederhana, dan terkadang interpretasi dapat lebih subjektif. Oleh karena itu, penting juga untuk mempertimbangkan konteks data dan tujuan klastering Anda sebelum memutuskan jumlah klaster yang optimal.

Sedangkan untuk pertanyaan menghitung Silhouette Score tanpa membuat visualisasi dengan menggunakan kode berikut:

```
library(eclust)

# Menjalankan k-means clustering
set.seed(100)
pca_km <- eclust(df_pca, stand = TRUE, FUNcluster = "kmeans", k = 4, graph = FALSE)  # Mengatur graph menjadi FALSE

# Menghitung Silhouette Score
silhouette_score <- silhouette(pca_km$cluster, dist(df_pca))
mean_silhouette <- mean(silhouette_score[, "sil_width"])

# Tampilkan nilai Silhouette Score
cat("Average Silhouette Score:", mean_silhouette, "\n")
```

**Reply by joejnt** [reply]
Dear @ramdhinafinita

Terima kasih untuk sarannya, tambahan satu lagi. Saya membuat model hierarchical clustering dengan perhitungan jarak menggunakan methode manhattaan dengan linkage methode menggunakan ward.D2 dengan syntax sebagai berikut:

df_dist_ma <- dist(x = df_pca, method = "manhattan")
ma_ward <- hclust(d = df_dist_ma, method = "ward.D2")
ward_clust <- cutree(ma_ward, k = 4)
ward_coph <- cophenetic(ma_ward)
cor(ward_coph, df_dist_ma)

untuk memvalidasi hasil cluster saya menggunakan library clValid() dengan syntax sebagai berikut:

internal <- clValid(df_pca, nClust = 3:5,
                    clMethods = "agnes",
                    validation = "internal",
                    metric = "manhattan",
                    method = "ward",
                    maxitems = nrow(df_pca))
summary(internal)

stability <- clValid(df_pca, nClust = 3:5,
                    clMethods = "agnes",
                    validation = "stability",
                    metric = "manhattan",
                    method = "ward",
                    maxitems = nrow(df_pca))
optimalScores(stability)

Namun masalah yang sama, saya temui adalah komputasi yang cukup lama. Apakah ada cara lain untuk memvalidasi model selain dari cara yang saya gunakan. Atau apakah ada kemungkinan salah syntax ya.

Mohon saran dan arahannya.

Best regards,
JT

**Comment by ramdhinafinita** [comment]
Dear @joejnt 

Pertanyaan Anda telah kami terima dan sedang kami usahakan terkait solusinya. Mohon kesediaannya untuk menunggu, terima kasih.

Best regards,
Team Algoritma

**Comment by ramdhinafinita** [comment]
Dear @joejnt 

Terimakasih atas perhatiannya sudah menunggu balasan dari kami. Berdasarkan kode yang Anda berikan, tidak terlihat ada kode yang keliru dan sdah benar. Hanya saja komputasi sangat lama memang dikarenakan Anda menggunakan Hierarchical Clustering dan berpengaruh juga pada cpu yang digunakan, sehingga kami bisa menyarankan Bapak untuk meminjam GPU dari Kaggle, dengan menjalankan modelnya melalui kaggle. 
Saran lain, bapak bisa lakukan perhitungan jaraknya jangan manhattan

Referensi: https://online.stat.psu.edu/stat555/node/85/

**Reply by joejnt** [reply]
Dear @ramdhinafinita 

Noted, nanti akan saya coba ganti untuk perhitungan jaraknya dg metode yang lain.

Terima kasih untuk inside nya.

Best regards,
JT

**Reply by ramdhinafinita** [reply, marked as answer]
Sama-sama @joejnt , jika jawaban dari pertanyaan yang diajukan bermanfaat dan dapat  diimplementasi silahkan lakukan "**Mark as answer"** pada jawaban ya pak, terimakasih
